src/dbnode/persist/fs/seek_manager.go

```diff
@@ -346,56 +383,50 @@ func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentD
-	for i, compareSeeker := range seekers.active.seekers {
-		if seeker == compareSeeker.seeker {
-			compareSeeker.isBorrowed = false
-			seekers.active.seekers[i] = compareSeeker
-			return true, nil
-		}
+	if m.markBorrowedSeekerAsReturned(seekers.active, seeker) {
+		// We can return right away if we've returned an active seeker.
+		return true, nil
 	}
 
 	// If no match was found in the active seekers, it's possible that an inactive seeker is being returned.
-	for i, compareSeeker := range seekers.inactive.seekers {
-		if seeker == compareSeeker.seeker {
-			compareSeeker.isBorrowed = false
-			seekers.inactive.seekers[i] = compareSeeker
-
-			// The goroutine that returns the last outstanding inactive seeker is responsible for notifying any
-			// goroutines waiting for all inactive seekers to be returned and clearing out the inactive seekers
-			// state entirely.
-			allAreReturned := true
-			for _, inactiveSeeker := range seekers.inactive.seekers {
-				if inactiveSeeker.isBorrowed {
-					allAreReturned = false
-					break
-				}
-			}
+	for i, inactiveSeeker := range seekers.inactive {
+		// Mark borrowed seeker as returned if any.
+		// We don't care about the return value here as we must continue the logic below.
+		m.markBorrowedSeekerAsReturned(inactiveSeeker, seeker)
+
+		// The goroutine that returns the last outstanding inactive seeker is responsible for notifying any
+		// goroutines waiting for all inactive seekers to be returned and clearing out the inactive seekers
+		// state entirely.
+		if inactiveSeeker.anyBorrowed() {
+			return true, nil
+		}
 
-			if !allAreReturned {
-				return true, nil
-			}
+		// If all borrowable seekers have been returned then we can safely close this inactive seeker
+		// and remove it from the list of inactive seekers.
+		seekers.removeInactiveAtIndex(i)
+		err := inactiveSeeker.close()
+		if inactiveSeeker.wg != nil {
+			// Signal completion regardless of any errors encountered while closing.
+			inactiveSeeker.wg.Done()
+		}
 
-			// All the inactive seekers have been returned so it's safe to signal and clear them out.
-			multiErr := xerrors.NewMultiError()
-			for _, inactiveSeeker := range seekers.inactive.seekers {
-				multiErr = multiErr.Add(inactiveSeeker.seeker.Close())
-			}
+		return true, err
+	}
 
-			// Clear out inactive state.
-			allInactiveSeekersClosedWg := seekers.inactive.wg
-			seekers.inactive = seekersAndBloom{}
-			if allInactiveSeekersClosedWg != nil {
-				// Signal completion regardless of any errors encountered while closing.
-				allInactiveSeekersClosedWg.Done()
-			}
+	return false, nil
+}
 
-			return true, multiErr.FinalError()
+func (m *seekerManager) markBorrowedSeekerAsReturned(seekers seekersAndBloom, seeker ConcurrentDataFileSetSeeker) bool {
+	for i, compareSeeker := range seekers.seekers {
+		if seeker == compareSeeker.seeker {
+			compareSeeker.isBorrowed = false
+			seekers.seekers[i] = compareSeeker
```

Comment:
Is this just a no-op? Aren't we for looping over the `seekers.seekers`?

Reply:
Ah, needed to pass in a pointer here after I refactored it.